enterprise/frontend/src/embedding-sdk/store/reducer.ts

```diff
@@ -0,0 +1,66 @@
+import { createAsyncThunk, createReducer } from "@reduxjs/toolkit";
+import type {
+  EmbeddingSessionTokenState,
+  SdkState,
+} from "embedding-sdk/store/types";
+
+const initialState: EmbeddingSessionTokenState = {
+  token: null,
+  loading: false,
+  error: null,
+};
+
+export const getSessionTokenState = (state: SdkState) =>
+  state.embeddingSessionToken;
+
+const GET_OR_REFRESH_SESSION =
+  "metabase-enterprise/embeddingSessionToken/GET_OR_REFRESH_SESSION";
+
+export const getOrRefreshSession = createAsyncThunk(
+  GET_OR_REFRESH_SESSION,
+  async (url: string, { dispatch, getState }) => {
+    const state = getSessionTokenState(getState());
+    const token = state?.token;
+
+    const isTokenValid = token && token.exp * 1000 >= Date.now();
+
+    if (!state.loading && !isTokenValid) {
```

Comment:
Would it be easier to understand if we do it like this?
```ts
if (isTokenValid) {
  return token
}

return dispatch(refreshTokenAsync(url))
```

It feels a little off trying to understand the condition when we should refresh the token. e.g. what if token is invalid, but the state is loading, in that case we're still sending `token`. Maybe that's impossible depends on how we construct the state, but it's a bit confusing as is.

Reply:
Yeah that's definitely better